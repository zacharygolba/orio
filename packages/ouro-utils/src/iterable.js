// @flow

type Target = { [key: string]: mixed }
type IterFn<+T> = () => Iterator<T>

function extractIteratorFn<T: Target>(target: T): ?IterFn<*> {
  const value = target[Symbol.iterator]

  if (typeof value === 'function') {
    return value
  }

  return undefined
}

export function isIterableObject(target: mixed): boolean %checks {
  return (
    target != null &&
    typeof target === 'object' &&
    typeof extractIteratorFn(target) === 'function'
  )
}

export function intoIterator(target: mixed): ?Iterator<*> {
  if (target != null && typeof target === 'object') {
    const iterFn = extractIteratorFn(target)

    if (typeof iterFn === 'function') {
      return iterFn.call(target)
    }
  }

  return undefined
}
