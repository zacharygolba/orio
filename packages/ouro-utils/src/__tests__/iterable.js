// @flow

import { isIterableObject, intoIterator } from '../iterable'

const customIterator = Object.freeze({
  next: () => ({
    done: true,
    value: undefined,
  }),
})

const customIterable = Object.freeze({
  [Symbol.iterator]: () => customIterator,
})

test('.isIterableObject()', () => {
  expect(isIterableObject(new Map())).toBe(true)
  expect(isIterableObject(new Set())).toBe(true)
  expect(isIterableObject(customIterable)).toBe(true)

  expect(isIterableObject({})).toBe(false)
})

test('.intoIterator()', () => {
  expect(intoIterator(new Map())).toMatchSnapshot()
  expect(intoIterator(new Set())).toMatchSnapshot()
  expect(intoIterator(customIterable)).toBe(customIterator)
})
