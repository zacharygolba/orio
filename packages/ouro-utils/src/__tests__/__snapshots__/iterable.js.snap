// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.intoAsyncIterator() 1`] = `
Object {
  "next": [Function],
}
`;

exports[`.intoAsyncIterator() 2`] = `
Object {
  "done": false,
  "value": "test",
}
`;

exports[`.intoIterator() 1`] = `Object {}`;

exports[`.intoIterator() 2`] = `Object {}`;
